### PR TITLE
[Broker] Fix call sync method in async rest api for internalGetMessageById.

### PR DESCRIPTION
**Pull request by Technoboy-**
### Motivation
Avoid call sync method in async rest API for PersistentTopicsBase#internalGetMessageById.

### Documentation
  
- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -2383,42 +2383,55 @@ private PositionImpl calculatePositionAckSet(boolean isExcluded, int batchSize,
...
+                                                            entry.release();
+                                                        }
+                                                    }
+                                                }
+                                            }, null);
+                                });
+                    });
+                }).exceptionally(ex -> {
+                    Throwable cause = ex.getCause();
+                    if (cause instanceof NullPointerException) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2426]: it's weird. why we catch NPE here ? 

I mean, if somewhere gets NPE, we need to check there. Shouldn't it be catching exceptions?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -2383,42 +2383,55 @@ private PositionImpl calculatePositionAckSet(boolean isExcluded, int batchSize,
...
+                                                        if (entry != null) {
+                                                            entry.release();
+                                                        }
+                                                    }
+                                                }
+                                            }, null);
+                                });
+                    });
+                }).exceptionally(ex -> {
+                    Throwable cause = ex.getCause();
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2425]: well, i think we probable missing to handle ``redirect RestException``.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2426]: Yes, maybe we will fix this in another new pr.  If fixed here, will be confused.

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2426]: Agree with you.

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2426]: Please drop this check.
It is the right time to get rid of this tech debt.

If someone sees the NPE we can fix it.
If we keep this now I bet no one will take care.

It is also possible that that NPE was due to some old bug


I would like to see you dropping this line

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: Done, This is involved by [6331](https://github.com/apache/pulsar/pull/6331). Seem that we can drop this line directly.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -2383,42 +2383,51 @@ private PositionImpl calculatePositionAckSet(boolean isExcluded, int batchSize,
...
-        } catch (NullPointerException npe) {
-            asyncResponse.resume(new RestException(Status.NOT_FOUND, "Message not found"));
-        } catch (Exception exception) {
-            log.error("[{}] Failed to get message with ledgerId {} entryId {} from {}",
-                    clientAppId(), ledgerId, entryId, topicName, exception);
-            asyncResponse.resume(new RestException(exception));
-        }
+                    return ret.thenCompose(ignore -> {
+                        return getTopicReferenceAsync(topicName)
+                                .thenAccept(topic -> {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2398]: Do we need to check if the topic is null?

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2398]: we already checked that the topic is here with validateTopicOwnershipAsync
I am not sure it is needed,
in the old code we did not check for nulls

**Review by eolivelli — APPROVED** [review]
+1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 3 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2398]: @codelipenghui If topic does not exist, getTopicReferenceAsync will throw exception.
And I have updated the logic. Thanks for the reminder.

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: > we already checked that the topic is here with validateTopicOwnershipAsync I am not sure it is needed, in the old code we did not check for nulls

getTopicReferenceAsync has checked and I have updated the logic here

**Pull request merged**